rig/machine_control/machine_controller.py

```diff
@@ -0,0 +1,382 @@
+        ----------
+        led : int
+            Number of the LED to set the state of (0-3)
+        """
+        arg1 = int(consts.LEDAction.off) << (led * 2)
+        self.send_scp(x, y, 0, SCPCommands.led, arg1=arg1, expected_args=0)
+
+    @ContextMixin.use_contextual_arguments
+    def led_toggle(self, led, x=Required, y=Required):
+        """Toggle the state of an LED.
+
+        Parameters
+        ----------
+        led : int
+            Number of the LED to set the state of (0-3)
+        """
+        arg1 = int(consts.LEDAction.toggle) << (led * 2)
+        self.send_scp(x, y, 0, SCPCommands.led, arg1=arg1, expected_args=0)
+
+    @ContextMixin.use_contextual_arguments
+    def sdram_alloc(self, size, tag=0, x=Required, y=Required,
+                    app_id=Required):
+        """Allocate a region of SDRAM for the given application.
+
+        Parameters
+        ----------
+        size : int
+            Number of bytes to attempt to allocate in SDRAM.
+        tag : int
+            8-bit tag that can be used identify the region of memory later.  If
+            `0` then no tag is applied.
+
+        Returns
+        -------
+        int
+            Address of the start of the region.
+
+        Raises
+        ------
+        SpiNNakerMemoryError
+            If the memory cannot be allocated, or the tag is already taken or
+            invalid.
+        """
+        assert 0 <= tag < 256
+
+        # Construct arg1 (op_code << 8) | app_id
+        arg1 = (consts.AllocOperations.alloc_sdram << 8 |
+                (self.app_id if app_id is None else app_id))
+
+        # Send the packet and retrieve the address
+        rv = self.send_scp(x, y, 0, SCPCommands.alloc_free, arg1, size, tag)
+        if rv.arg1 == 0:
+            # Allocation failed
+            raise SpiNNakerMemoryError(size, x, y)
+        return rv.arg1
+
+CoreInfo = collections.namedtuple(
+    'CoreInfo', "p2p_address physical_cpu virt_cpu version "
+                "buffer_size build_date version_string")
+
+
+class IPTag(collections.namedtuple("IPTag",
+                                   "addr max port timeout flags count rx_port "
+                                   "spin_addr spin_port")):
+    """An IPTag as read from a SpiNNaker machine."""
+    @classmethod
+    def from_bytestring(cls, bytestring):
+        (ip, max, port, timeout, flags, count, rx_port, spin_addr,
+         spin_port) = struct.unpack("4s 6s 3H I 2H B", bytestring[:25])
+        # Convert the IP address into a string, otherwise save
+        ip_addr = '.'.join(str(x) for x in struct.unpack("4B", ip))
+
+        return cls(ip_addr, max, port, timeout, flags, count, rx_port,
+                   spin_addr, spin_port)
+
+
+# Dictionary of (address % 4, n_bytes % 4) to data type
+address_length_dtype = {
+    (i, j): (DataType.word if (i == j == 0) else
+             (DataType.short if (i % 2 == j % 2 == 0) else
+              DataType.byte)) for i in range(4) for j in range(4)
```

Comment:
Breaking the "for" and aligning with the `(i, j):` would make this far more obvious (to me, anyway) -- it took me a good while to spot!